# Rebased directly onto a2flo/floor_llvm 4b173e3db383ec7164afb11ee7dc86609c88f052.
# Provenance and conflict-resolution notes are recorded in manifest.json rebaseProvenance.
--- a/clang/lib/CodeGen/BackendUtil.cpp
+++ b/clang/lib/CodeGen/BackendUtil.cpp
@@ -46,13 +46,19 @@
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Support/MemoryBuffer.h"
 #include "llvm/Support/PrettyStackTrace.h"
+#ifdef CLANG_FLOOR_ENABLE_SPIRV
 #include "llvm/../../projects/spirv/include/LLVMSPIRVLib.h"
+#endif
 #include "llvm/Support/TimeProfiler.h"
 #include "llvm/Support/Timer.h"
 #include "llvm/Support/ToolOutputFile.h"
 #include "llvm/Support/raw_ostream.h"
+#ifdef CLANG_FLOOR_ENABLE_SPIRV
 #include "llvm/../../projects/spirv/lib/SPIRV/SPIRVWriterPass.h"
+#endif
+#ifdef CLANG_FLOOR_ENABLE_SPIRV
 #include "llvm/../../projects/spirv/lib/SPIRV/SPIRVContainerWriterPass.h"
+#endif
 #include "llvm/Target/TargetMachine.h"
 #include "llvm/Target/TargetOptions.h"
 #include "llvm/Transforms/Coroutines.h"
@@ -1177,11 +1183,21 @@
     break;
 
   case Backend_EmitSPIRV:
+#ifdef CLANG_FLOOR_ENABLE_SPIRV
     PerModulePasses.add(createSPIRVWriterPass(*OS, CodeGenOpts.SPIRVVersion));
+#else
+    TheModule->getContext().emitError("SPIR-V emission is unavailable in this AIR-only build");
+    return;
+#endif
     break;
 
   case Backend_EmitSPIRVContainer:
+#ifdef CLANG_FLOOR_ENABLE_SPIRV
     PerModulePasses.add(createSPIRVContainerWriterPass(*OS));
+#else
+    TheModule->getContext().emitError("SPIR-V emission is unavailable in this AIR-only build");
+    return;
+#endif
     break;
 
   case Backend_EmitMetalLib:
@@ -1641,11 +1657,21 @@
     break;
 
   case Backend_EmitSPIRV:
+#ifdef CLANG_FLOOR_ENABLE_SPIRV
     MPM.addPass(SPIRVWriterPass(*OS));
+#else
+    TheModule->getContext().emitError("SPIR-V emission is unavailable in this AIR-only build");
+    return;
+#endif
     break;
 
   case Backend_EmitSPIRVContainer:
+#ifdef CLANG_FLOOR_ENABLE_SPIRV
     MPM.addPass(SPIRVContainerWriterPass(*OS));
+#else
+    TheModule->getContext().emitError("SPIR-V emission is unavailable in this AIR-only build");
+    return;
+#endif
     break;
 
   case Backend_EmitMetalLib:
--- a/clang/lib/CodeGen/CMakeLists.txt
+++ b/clang/lib/CodeGen/CMakeLists.txt
@@ -1,3 +1,9 @@
+# The optional floor SPIR-V project is not shipped in standalone AIR builds.
+option(CLANG_FLOOR_ENABLE_SPIRV "Enable the separately supplied floor SPIR-V writers" OFF)
+if(CLANG_FLOOR_ENABLE_SPIRV)
+  add_definitions(-DCLANG_FLOOR_ENABLE_SPIRV=1)
+endif()
+
 set(LLVM_LINK_COMPONENTS
   Analysis
   BitReader
@@ -26,11 +32,15 @@
   ProfileData
   Remarks
   ScalarOpts
-  SPIRVLib
+  SelectionDAG
   Support
   Target
   TransformUtils
   )
+
+if(CLANG_FLOOR_ENABLE_SPIRV)
+  list(APPEND LLVM_LINK_COMPONENTS SPIRVLib)
+endif()
 
 add_clang_library(clangCodeGen
   BackendUtil.cpp
